### ISSUE TITLE
Memberlist not propagating to all nodes when a swarm manager is down

### ISSUE THREAD

**Issue by mefyl** [docker/libnetwork#1734]
On a 10 nodes swarm with one manager down (mefyl-swarm-1):

```
root@mefyl-swarm-0:~# docker node ls
ID                           HOSTNAME       STATUS  AVAILABILITY  MANAGER STATUS
0fxz3nzxudjqgo0bbut0rf5ns    mefyl-swarm-9  Ready   Active
6puobklis16gi3reelmymug4b *  mefyl-swarm-0  Ready   Active        Leader
auz85qw028dl7zdvvfdadktys    mefyl-swarm-6  Ready   Active
dcplnks00s7bgrpgv5vcjcqtn    mefyl-swarm-3  Ready   Active
enimyxgw3c14fnq3o7tgik6h7    mefyl-swarm-8  Ready   Active
mncaaomyobbgen1pay24uxc9i    mefyl-swarm-4  Ready   Active
nn9b5zp87fyknv4fzhq0o3hag    mefyl-swarm-2  Ready   Active        Reachable
nzwrquw11f3owuvejsose9jhc    mefyl-swarm-7  Ready   Active
rnfpnb4w44pvrrlzom415lhv4    mefyl-swarm-5  Ready   Active
ycppe6vu7jovw43atf0nofikt    mefyl-swarm-1  Down    Active        Unreachable
```

I create an `infinit` service, part of an `infinit` overlay network. It seems that my infinit process is not able
to reach some other nodes : 3 nodes are entirely isolated, while the 6 other are able to see each others (with 1 node down this account for the 10 of them).

Inspecting the network on all hosts reflects this 6-1-1-1 situation:

```
$pssh -i -Hmefyl-swarm-{0,1,2,3,4,5,6,7,8,9} "docker network inspect infinit | jq '.[].Peers'"
[1] 18:28:20 [SUCCESS] mefyl-swarm-3
[
  {
    "Name": "mefyl-swarm-3-18b7dca6d048",
    "IP": "207.154.216.219"
  }
]
[2] 18:28:20 [SUCCESS] mefyl-swarm-5
[
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  },
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  }
]
[3] 18:28:20 [SUCCESS] mefyl-swarm-0
[
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  },
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  },
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  }
]
[4] 18:28:20 [SUCCESS] mefyl-swarm-8
[
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  },
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  }
]
[5] 18:28:20 [SUCCESS] mefyl-swarm-2
[
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  },
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  },
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  }
]
[6] 18:28:20 [SUCCESS] mefyl-swarm-6
[
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  },
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  }
]
[7] 18:28:20 [SUCCESS] mefyl-swarm-9
[
  {
    "Name": "mefyl-swarm-9-6959cf578397",
    "IP": "207.154.239.119"
  },
  {
    "Name": "mefyl-swarm-0-e9e9f023ec34",
    "IP": "207.154.206.246"
  },
  {
    "Name": "mefyl-swarm-8-ba9795b23f80",
    "IP": "207.154.239.86"
  },
  {
    "Name": "mefyl-swarm-6-6953774050fb",
    "IP": "207.154.215.254"
  },
  {
    "Name": "mefyl-swarm-5-8cf5bb13104b",
    "IP": "207.154.215.162"
  },
  {
    "Name": "mefyl-swarm-2-937701d1f653",
    "IP": "207.154.195.15"
  }
]
[8] 18:28:20 [SUCCESS] mefyl-swarm-1
Stderr: Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?
[9] 18:28:20 [SUCCESS] mefyl-swarm-7
[
  {
    "Name": "mefyl-swarm-7-ab3114826b2f",
    "IP": "207.154.224.151"
  }
]
[10] 18:28:21 [SUCCESS] mefyl-swarm-4
[
  {
    "Name": "mefyl-swarm-4-81704d59f248",
    "IP": "207.154.217.182"
  }
]
```

After some debugging with @mavenugo, it seems that `mefyl-swarm-1` being down is likely the cause of this partition:

```
Apr 27 09:34:48 mefyl-swarm-7 dockerd[23816]: time="2017-04-27T09:34:48.278179898Z" level=debug msg="memberlist: Failed to join 207.154.193.246: dial tcp 207.154.193.246:7946: getsockopt: connection refused" 
Apr 27 09:34:48 mefyl-swarm-7 dockerd[23816]: time="2017-04-27T09:34:48.278709834Z" level=error msg="Failed to join memberlist [207.154.193.246] on retry: 1 error(s) occurred:\n\n* Failed to join 207.154.193.246: dial tcp 2
```

Docker version:

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by fcrisciani** [comment]
So far the current flow of the events identified this morning from the setup are the following:
1) swarm cluster started with 1 manager
2) 7 workers added to the swarm cluster
3) 2 manager added to the swarm cluster
== Issue starting ==
4) Apr 18th manager 2 reach out of memory issue and panics
During this event most likely it corrups also the RAFT db
5) systemd tries to restart the daemon but because of the RAFT being corrupted the manager does not restart properly. 
6) After a certain number of retries systemd rightfully gives up.
====
More on the libnetwork side there are 3 workers that from a network db point of view are isolated (are not part of the cluster and are not able to join it anymore)

On the setup the logs unfortunately are all rotated so the least recent timeline that we have is from Apr 27th.

Looking at that state is clear that the network db is in a retry loop for joining the cluster, but the interesting thing is that all these 3 nodes are trying to join the cluster connecting to 207.154.193.246

Now this is actually the IP of the swarm manager that paniced and is gone forever, so there is no way for the gossip to establish again the adjacency leaving these 3 nodes isolated forever.

The root cause of this is in the method ``GetRemoteAddress()`` that is used to seed the IP address to the network db. This function just returns a random manager IP with no knowledge if the manager is actually available or not.

Will start working on a patch for this case, and changing the ``GetRemoteAddress()`` to return the list of managers instead of choosing one. The network db is already able to handle a list of IP to use to establish the initial adjacencies.

**Comment by fcrisciani** [comment]
All the commits are in @mavenugo you can go ahead and close this one then

**Comment by aboch** [comment]
Fixed by #1737
